src/sorting/counting_sort.rs

```diff
@@ -1,29 +1,64 @@
-pub fn counting_sort(array: &[usize]) -> Vec<usize> {
-    let max = array.iter().max().unwrap_or(&0);
+use std::collections::HashMap;
 
-    let mut count = vec![0; max + 1];
-    let mut output = vec![0; array.len()];
+pub fn counting_sort<T: Keyed + Ord + Copy>(array: &[T]) -> Vec<T> {
+    let mut new_hashmap: HashMap<usize, &T> = HashMap::new();
 
-    for &element in array.iter() {
+    for element in array {
+        new_hashmap.insert(element.key(), element);
+    }
+
+    let mut usize_array: Vec<usize> = Vec::new();
+
+    for element in array {
+        usize_array.push(element.key());
+    }
+
+    let max = usize_array.iter().max().unwrap_or(&0);
+    let mut count: Vec<usize> = vec![0; max + 1];
+    let mut output: Vec<T> = vec![array[0]; array.len()];
```

Comment:
If we do this, we have to check that the array isn't empty because then it would panic.

Reply:
We could add a constraint of `T` having a `Default` implementation, but I leave that decision to you. 😁 

That would leave us with:

```rust
let mut output = vec![T::default(); array.len()];
```

Reply:
This is finding for me:
```rust
T::default();
```
Thanks